Sprint_3

nikkotin342 opened the pull request.

Commit:
added tests
added info in README.md
added .idea in .gitignore

irisqul commented on tests/test_from_lk_to_construction.py:
```diff
@@ -0,0 +1,33 @@
+from selenium.webdriver.support import expected_conditions
+from selenium.webdriver.support.wait import WebDriverWait
+
+driver = webdriver.Chrome()
```
Нужно исправить: здесь и далее по всем тестам:

1. В каждом тестовом модуле необходимо добавить тестовые методы, а не просто исполняемый код
2. Тестовые методы внутри одного тестового модуля объединяют в тестовые классы.
3. Один тестовый класс (соответственно один тестовый модуль) на несколько методов, проверяющих схожую функциональность
4. Инстанс драйвера необходимо получать и закрывать в фикстуре
5. Все локаторы необходимо хранить отдельно от тестов в модуле locators
6. Локаторы необходимо привязываться к атрибутам конкретного элемента. Ходить от root, использовать порядковый номер элемента в dom и т.д. является весьма нестабильным способом поиска

nikkotin342 replied:
Добрый вечер! Внес все исправления, только хотел уточнить что-то нужно изменить в  описании к заданию в файле README.md?

Commit:
Изменения в рамках ревью.
Добавлены классы и функции
Добавлена фикстура
Добавлен файл с локаторами
Изменены локаторы на обобщенный формат
Тесты приведены к запуску через pytest

nikkotin342 commented:
Не понял из объяснений в практикуме, как надо делать правильно с ревью пул реквеста, поэтому запушил новый коммит сюда, если необходимо старый пул реквест удалить и залить новый, я сделаю без проблем, просто думаю что с закрытием пул реквеста удалятся комментарии для ревью)

irisqul commented:
> Не понял из объяснений в практикуме, как надо делать правильно с ревью пул реквеста, поэтому запушил новый коммит сюда, если необходимо старый пул реквест удалить и залить новый, я сделаю без проблем, просто думаю что с закрытием пул реквеста удалятся комментарии для ревью)

да, всё верно - продолжаем работать в том пулл реквесте где проведено ревью. так мы видим всю историю комментов, можем отмечать что уже поправлено и т.п

irisqul commented on tests/test_from_lk_to_construction.py:
```diff
@@ -0,0 +1,27 @@
+        driver.find_element(*TestLocators.SEARCH_ENT_BUTTON).click()
+        driver.find_element(*TestLocators.SEARCH_BUTTON_LK).click()
+        driver.find_element(*TestLocators.SEARCH_BUTTON_CONSTRUCTION).click()
+        WebDriverWait(driver, 3).until(expected_conditions.visibility_of_element_located(TestLocators.SEARCH_CONTRUCTION_FIELD))
```
Нужно исправить: нужны еще ассерты - каждый тест должен заканчиваться проверкой.

irisqul commented on tests/test_from_lk_to_construction.py:
```diff
@@ -0,0 +1,27 @@
+class TestFromLkToConstruction:
+
+    def test_from_lk_to_construction_button(self, driver):
+        driver.get("https://stellarburgers.nomoreparties.site/")
```
Нужно исправить: вынести все урлы в отдельный файл и переиспользовать их. Тогда, если нужно будет тестировать не продакшн нам нужно будет изменить урлы только в одном файле, а не искать их по всему проекту

irisqul commented on tests/locators.py:
```diff
@@ -0,0 +1,27 @@
+from selenium.webdriver.common.by import By
+
+class TestLocators:
+    SEARCH_BUTTON_ENTER = [By.XPATH, '//*[contains(@class,"button_button_size_large") and text()="Войти в аккаунт"]']# Кнопка "Войти в аккаунт"
```
Отличные локаторы, молодец!

irisqul commented on tests/locators.py:
```diff
@@ -0,0 +1,27 @@
+    SEARCH_BUTTON_SAUCES = [By.XPATH, '//*[contains(@class, "text_type_main-default") and text() = "Соусы"]']# Кнопка "Соусы" в конструкторе
+    SEARCH_BUTTON_STAFFS = [By.XPATH, '//*[contains(@class, "text_type_main-default") and text() = "Начинки"]']# Кнопка "Начинки" в конструкторе
+    SEARCH_BUTTON_BREADS = [By.XPATH, '//*[contains(@class, "text_type_main-default") and text() = "Булки"]']# Кнопка "Булки" в конструкторе
+    SEARCH_FIELD_SAUCES = [By.XPATH, '//*[contains(@class, "text_type_main-medium") and text() = "Соусы"]']# Заголовок "Соусы" в поле с соусами в ингредиентах конструктора
```
при активации табов нужно смотреть класс current , пожалуйста обнови тут селекторы

nikkotin342 replied:
Не совсем понял, что требуется сделать( Нужно изменить имя локаторов по соусам, булкам и начинкам, чтобы долго не искать нажатием таб что именно нужно а найти быстрее, верно понял?

irisqul commented on tests/test_registration.py:
```diff
@@ -0,0 +1,23 @@
+        driver.find_element(*TestLocators.SEARCH_FIELD_EMAIL).send_keys('bakki_98@mail.ru')
+        driver.find_element(*TestLocators.SEARCH_FIELD_PASS).send_keys('12345')
+        driver.find_element(*TestLocators.SEARCH_BUTTON_REGSAVE).click()
+        assert driver.find_element(*TestLocators.SEARCH_INCORRECT_PASS).text == 'Некорректный пароль'
```
вот тут хороший тест, с ассертом

irisqul commented on tests/test_registration.py:
```diff
@@ -0,0 +1,23 @@
+        driver.find_element(*TestLocators.SEARCH_BUTTON_ENTER).click()
+        driver.find_element(*TestLocators.SEARCH_BUTTON_REG).click()
+        driver.find_element(*TestLocators.SEARCH_FIELD_NAME).send_keys('Сашок')
+        driver.find_element(*TestLocators.SEARCH_FIELD_EMAIL).send_keys('bakki_98@mail.ru')
```
Нужно исправить: одноразовый тест. Чтобы проходила регистрация нового пользователя email каждый тест должен быть новый. Лучшее место для генерации тестовых данных - это фикстура

Commit:
Изменения в рамках ревью №2
Добавлены ассерты в тесты
Добавлены новые локаторы
Добавлены и подключены новые фикстуры

nikkotin342 commented:
В рамках 2 ревью внес правки согласно комментариям выше) В test_select_element_of_construction добавил ассерт после каждого выбора элемента конструктора, это верно или нужно общий ассерт сделать?

irisqul commented on tests/locators.py:
```diff
@@ -0,0 +1,35 @@
+    SEARCH_FIELD_SAUCES = [By.XPATH, '//*[contains(@class, "text_type_main-medium") and text() = "Соусы"]']# Заголовок "Соусы" в поле с соусами в ингредиентах конструктора
+    SEARCH_FIELD_STAFFS = [By.XPATH, '//*[contains(@class, "text_type_main-medium") and text() = "Начинки"]']# Заголовок "Начинки" в поле с соусами в ингредиентах конструктора
+    SEARCH_FIELD_BREADS = [By.XPATH, '//*[contains(@class, "text_type_main-medium") and text() = "Булки"]']# Заголовок "Булки" в поле с соусами в ингредиентах конструктора
+    GET_MAIN_URL = 'https://stellarburgers.nomoreparties.site/'# URL стартовой страницы сайта Stellar Burgers
```
Нужно исправить - в этом файле только локаторы, урлы нужно положить в отдельный файл. Это очень разные данные и используем мы их по разному, поэтому нельзя перемешивать

irisqul commented on tests/locators.py:
```diff
@@ -0,0 +1,35 @@
+    SEARCH_BUTTON_REG = [By.XPATH, '//*[contains(@class,"Auth_link")]']# Кнопка "Зарегистрироваться"
+    SEARCH_FIELD_NAME = [By.XPATH, '//*[contains(@class, "input__textfield") and @value = ""]']# Заполнить поле для Имени
+    SEARCH_FIELD_EMAIL = [By.XPATH, '//*[contains(@class, "input__textfield") and @value = ""]']# Заполнить поле для Емаил
+    SEARCH_FIELD_PASS = [By.XPATH, '//*[contains(@class, "input__textfield") and @value = ""]']# Заполнить поле для Пароля
```
вот тут не заметила сразу - три одинаковых селектора подряд, значит неверные они и не обеспечивают поиск конкретного элемента стабильно. каждый селектор и локатор должны быть уникальны, на одной странице уж точно

irisqul commented on tests/test_from_lk_to_construction.py:
```diff
@@ -0,0 +1,29 @@
+        driver.find_element(*TestLocators.SEARCH_BUTTON_LK).click()
+        driver.find_element(*TestLocators.SEARCH_BUTTON_CONSTRUCTION).click()
+        WebDriverWait(driver, 3).until(expected_conditions.visibility_of_element_located(TestLocators.SEARCH_CONTRUCTION_FIELD))
+        assert driver.find_element(*TestLocators.SEARCH_CONTRUCTION_HEADER).text == 'Соберите бургер'
```
Нужно исправить: мы проверяем не текст, а наличие элемента. Нужно перенести поиск элемента в ассерт, сейчас если есть баг то тест упадет не на ассерте, а на предыдущем шаге

irisqul commented on tests/test_registration.py:
```diff
@@ -0,0 +1,28 @@
+class TestRegistration:
+
+
+    def test_registration_success(self, driver, email, password_correct):
```
отлично, генерируются новые данные для каждой сессии

Commit:
Изменения в рамках ревью №3
Изменены ассерты в тесты
Добавлен файл help_for_tests.py с УРЛ сайта
Добавлена информация в README.md
